### DISCUSSION TITLE
Use of VS Code's Development Container

### DISCUSSION THREAD

**Discussion by astroesteban** [nasa/fprime#908] [Show and tell]
### About
Hey y'all. I've been going through the tutorials for F' using the Visual Studio Code's development container feature. It's nothing out of this world but I thought I'd share it in case any newbies like me are out there and just want a simple way to get up and running with F'.

You can find out more info about devcontainers [here](https://code.visualstudio.com/docs/remote/create-dev-container).
Here is a great YouTube playlist from Microsoft on development containers: https://youtu.be/61M2takIKl8

Basically they allow you to use a Docker container as your development environment. VS Code automatically recognizes the `.devcontainer` folder and mounts your project into the container in the background. It's great for newbies to get setup and running as well as creating a common development environment for everyone.

I based the devcontainer on the already existing `Dockerfile` so that there aren't any duplicate Dockerfiles floating around.

With this environment you can build F', run the HTML GUI, and I even went ahead and installed the autocompletion for the CLI tools.

Check it out here: https://github.com/astroesteban/fprime/tree/feature/add-devcontainer/.devcontainer

### Setup

In order to use dev containers you will need:

- [Docker](https://www.docker.com)
- [Visual Studio Code](https://code.visualstudio.com)
- [Remote Development Extension Pack](https://marketplace.visualstudio.com/items?itemName=ms-vscode-remote.vscode-remote-extensionpack)

Install Docker and Visual Studio Code then in VS Code install the Remote Development Extension Pack. The extension pack will allow you to develop using SSH or, for our purpose, a dev container.

Next, all you need to do is copy the _.devcontainer_ folder to your F' fork and then open the repo in VS Code.

Once you have the repo opened in VS Code you should see a pop-up in the bottom right:
<img width="587" alt="Screen Shot 2021-08-18 at 11 30 59 AM" src="https://user-images.githubusercontent.com/12473592/129936646-7959ec63-894f-44cc-b22e-78fa58659401.png">

If you don't see that you can click on the magenta icon in the bottom left:
<img width="407" alt="Screen Shot 2021-08-18 at 11 32 42 AM" src="https://user-images.githubusercontent.com/12473592/129936809-47e19d43-59c9-4077-bdfd-4b76dea91771.png">

and then click on "Reopen in Container"
<img width="623" alt="Screen Shot 2021-08-18 at 11 33 58 AM" src="https://user-images.githubusercontent.com/12473592/129936967-a92397b6-900e-42e1-aa94-4d566c78033b.png">

VS Code will build and start the container. Then it will mount your repo as a cached volume inside of the container so any changes made in the devcontainer are to your host repo.

You should end up with a bottom half that looks like this:
<img width="934" alt="Screen Shot 2021-08-18 at 11 37 21 AM" src="https://user-images.githubusercontent.com/12473592/129937455-4c680732-cf00-41f4-b96e-fac1d28b5fc4.png">

**Comment by Joshua-Anderson** [comment]
That's really cool! I usually spin up a vm for my linux testing but this could be a great alternative.

**Reply by astroesteban** [reply]
Spinning up a VM was my go-to but recently I've found that VS Code + DevContainers have been great at just quickly getting up and running with a project. Plus I don't need to create new SSH keys for every VM or copy them over. It's pretty cool stuff coming out of the VS Code team.

**Comment by astroesteban** [comment]
Updated the discussion with setup info in case anyone comes across this in the future and wants to copy it.

**Comment by ChangoBuitrago** [comment]
Hi @astroesteban, thanks for the above suggestion, looks great. Unfortunately, the check it out link is not working at the moment. It throughs a `404` error. It could just be a permissions issue (https://docs.github.com/en/enterprise-server@3.0/rest/overview/troubleshooting). 
TIA.

**Reply by astroesteban** [reply]
Hey @ChangoBuitrago my apologies for responding so late. I was out on vacation. You can find the devcontainer folder [here](https://github.com/astroesteban/fprime/tree/feat/devcontainer-demo/.devcontainer).

**Reply by ChangoBuitrago** [reply]
Hi @astroesteban. On the contrary, thank you for taking the time to reshare this. I'll give it a try ASAP and let you know. 
Hope you had a great holiday.
Cheers!

**Reply by astroesteban** [reply]
Thanks @ChangoBuitrago! I decided to create my own Dockerfile in .devcontainer instead of using the existing Dockerfile in F'. Having my own gives me greater control over what I install in my dev environment and I can also easily select between Ubuntu 20.04 or 18.04. Everything else is the same as the existing Dockerfile including the virtual environment setup and package installations.

**Reply by ChangoBuitrago** [reply]
You are welcome @astroesteban. I did the same but upgraded the existent `Dockerfile`. However, even though I managed to have the whole system up and running (exposing IPs, Ports, etc), I finally couldn't access the `GDS` flask app no matter what. This is the main reason why I got interested in your solution because I do need a working Fprime Docker-based version.
TIA, 
Regards

**Reply by Joshua-Anderson** [reply]
@ChangoBuitrago Did you try passing in the `--gui-addr 0.0.0.0` argument to the GDS? By default the GDS gui only listens on 127.0.0.1, which wouldn't be accessible outside the docker container.

**Reply by ChangoBuitrago** [reply]
Thanks, @Joshua-Anderson. I did, but I was kind of in a rush at that time, so I may have missed something. I will give it a try again next week and report back. 
Thanks for your help and interest.
